docs/design/cpsubsystem/02-group-availability-listener.md

```diff
@@ -0,0 +1,145 @@
+     *
+     * @param event CP group availability event
+     */
+    void availabilityDecreased(CPGroupAvailabilityEvent event);
+
+    /**
+     * Called when a CP group has lost its majority.
+     *
+     * @param event CP group availability event
+     */
+    void majorityLost(CPGroupAvailabilityEvent event);
+}
+```       
+
+```java 
+/**
+ * CPGroupAvailabilityEvent is published when a CP group's
+ * availability is decreased or it has lost the majority completely.
+ *
+ * @since 4.1
+ */
+public interface CPGroupAvailabilityEvent {
+
+    /**
+     * Returns the id of the related CP group.
+     *
+     * @return CP group id
+     */
+    CPGroupId getGroupId();
+
+    /**
+     * Returns the current members of the CP group.
+     *
+     * @return group members
+     */
+    Collection<CPMember> getGroupMembers();
+
+    /**
+     * Returns the unavailable members of the CP group.
+     *
+     * @return unavailable members
+     */
+    Collection<CPMember> getUnavailableMembers();
+
+    /**
+     * Returns the majority member count of the CP group.
+     * Simply it's {@code (memberCount/2) + 1}
+     *
+     * @return majority
+     */
+    int getMajority();
+
+    /**
+     * Returns whether this group has the majority of its members available or not.
+     *
+     * @return true if the group still has the majority, false otherwise
+     */
+    boolean isMajorityAvailable();
+
+    /**
+     * Returns whether this is the METADATA CP group or not.
+     *
+     * @return true if the group is METADATA group, false otherwise
+     */
+    boolean isMetadataGroup();
+}
+```
+
+## Technical Design
+
+CP group operations are performed by METADATA group. Because of this, events for group availability changes are published 
+by METADATA members.
+
+In general, availability decreases when a CP member becomes unreachable because of process crash, network partition, 
+out of memory etc. Tracking CP group availability is not as strict as CP membership changes. It relies on Hazelcast's 
+unreliable membership failure detectors. Once a member is declared as unavailable by the Hazelcast's failure detector, 
+that member is removed from the cluster. When a Hazelcast member is removed from the cluster, all METADATA group members 
+iterate over CP groups to find out the ones having a member with the same address as the removed member. If a group is 
+affected by the member's removal, then a new `CPGroupAvailabilityEvent` is published for that group. That group can be 
+METADATA group itself too.
+
+Since all available METADATA members produce the same `CPGroupAvailabilityEvent` for a specific member removal, 
+it's possible to receive multiple instances of the same event. To avoid this issue, a deduplication filter is applied 
+before invoking `CPGroupAvailabilityListener`. Events with the exact same signature in a _one minute_ period will be 
+discarded, only the first instance of them will be passed to the listeners. 
```

Comment:
How was this "one minute" decided? Based on what?

Reply:
I think one min is long enough window to receive events from all members. 

- If this value is too high, then we can skip some real new events. For instance, when persistence is enabled, a member can come and go a in a few minutes. 
- If it's too low, then we can show fire multiple events for a single event when a late event comes.

It it was 5mins, then you would ask why it is 5min. If I make it configurable, then I would need a default value. And how would user decide on a good value?

Reply:
Ok, sounds reasonable. I'm fine with the decision then as I understand the problems.